sdlib/d/sync/sharedlock.d

```diff
@@ -0,0 +1,558 @@
+		assert(lock.mutex.isHeld(), "Mutex not held while exclusively locked!");
+		state.store(1);
+		lock.exclusiveUnlock();
+
+		while (state.load() != 2) {
+			import sys.posix.sched;
+			sched_yield();
+		}
+
+		return null;
+	}
+
+	void* run2() {
+		while (state.load() != 1) {
+			import sys.posix.sched;
+			sched_yield();
+		}
+
+		assert(state.load() == 1);
+
+		lock.exclusiveLock();
+		assert(lock.count == SharedLock.Exclusive, "Invalid lock state!");
+		assert(lock.mutex.isHeld(), "Mutex not held while exclusively locked!");
+		state.store(2);
+		lock.exclusiveUnlock();
+
+		return null;
+	}
+
+	auto t1 = runThread(run1);
+	auto t2 = runThread(run2);
+
+	import core.stdc.pthread;
+	void* ret;
+	pthread_join(t1, &ret);
+	pthread_join(t2, &ret);
+
+	assert(!lock.mutex.isHeld(), "Mutex is held!");
+	assert(lock.count == 0, "Invalid lock state!");
+}
+
+unittest exclusiveAndSharedLock {
+	/**
+	 * Test the following scenarios:
+	 * 1. When a shared lock is held, exclusively locking waits.
+	 * 2. When a shared lock is held, and exclusive lock is sought, further
+	 *    shared locks should wait.
+	 * 3. When all shared locks are released, the exclusive lock is taken.
+	 * 4. When the exclusive lock is released, shared locks can again be taken.
+	 *
+	 * To accomplish this, we will start one exclusive thread and several
+	 * sharing threads. We will issue commands for each action, and then
+	 * wait for the desired state, then we measure the lock state and
+	 * ensure everything is correct.
+	 */
+	static runThread(void* delegate() dg) {
+		static struct Delegate {
+			void* ctx;
+			void* function(void*) fun;
+		}
+
+		auto x = *(cast(Delegate*) &dg);
+
+		import core.stdc.pthread;
+		pthread_t tid;
+		auto r = pthread_create(&tid, null, x.fun, x.ctx);
+		assert(r == 0, "Failed to create thread!");
+
+		return tid;
+	}
+
+	shared SharedLock lock;
+
+	import d.sync.atomic;
+	import sys.posix.sched;
+	shared Atomic!uint exclusiveLockState;
+	shared Atomic!uint sharedLockCount;
+	shared Atomic!uint attemptedSharedLockCount;
+	shared Atomic!uint desiredSharedLocks;
+
+	void* exclusiveLocker() {
+		while (true) {
+			uint cur = exclusiveLockState.load();
+			auto s = cur;
+			assert((s & 1) == 0);
+			while (s == cur) {
+				sched_yield();
+				s = exclusiveLockState.load();
+			}
+
+			scope(exit) exclusiveLockState.fetchAdd(1);
+
+			switch (s) {
+				case 1: // Lock
+					lock.exclusiveLock();
+					break;
+				case 3: // Unlock
+					lock.exclusiveUnlock();
+					break;
+				default: // Exit
+					return null;
+			}
+		}
+	}
+
+	void* sharedLocker() {
+		uint id = sharedLockCount.fetchAdd(1);
+		bool lockHeld = false;
+		while (true) {
+			uint cmd = desiredSharedLocks.load();
+			if (cmd == uint.max) {
+				return null;
+			}
+
+			if (cmd > id && !lockHeld) {
+				attemptedSharedLockCount.fetchAdd(1);
+				lock.sharedLock();
+				attemptedSharedLockCount.fetchSub(1);
+				lockHeld = true;
+				sharedLockCount.fetchAdd(1);
```

Comment:
It does look like sharedLockCount is used for two different things. If the other one is just to give the thread an id, you can pass the id for it in the closure, which would make everything nicer.

Reply:
I would have to allocate closures for this, I didn't want to do that.

I will add a dedicated counter for the id.